scripts/build-pack.sh

```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
+
+### This script is, in principle, similar to download-env.sh
+### Notable differences:
+### - download-env does not download everything, but depends on system variables
+### - download-env uses user's site-packages
+### - download-env installs LiteX submodules. This is not allowed here as
+###   conda-pack fails when it finds editable pip modules
+### - download-env downloads Zephyr SDK if needed. This script does not do it,
+###   but will when the SDK conda package is created
+
+SETUP_SRC="$(realpath ${BASH_SOURCE[0]})"
+SETUP_DIR="$(dirname "${SETUP_SRC}")"
+TOP_DIR="$(realpath "${SETUP_DIR}/..")"
+
+source "$SETUP_DIR/setup-helpers.sh"
+source "$SETUP_DIR/settings.sh"
+
+verify_system || exit 1
+
+set -e
+set -u
+
+# Ensure the build dir
+mkdir -p $BUILD_DIR
+
+echo ""
+echo "Initializing environment"
+echo "------------------------"
+# Install and setup conda for downloading packages
+export PATH=$CONDA_DIR/bin:$PATH:/sbin
+(
+	install_conda
+)
+
+echo ""
+echo "Preparing packages"
+echo "------------------"
+
+# this file will contain final env description
+ENV_FILE=$BUILD_DIR/environment.yml
+
+# this file contains packages which versions are not specified in settings.sh
+ENV_TEMPLATE=$SETUP_DIR/template-environment.yml
+
+# This is the name of the new environment from which the pack will be created.
+# Note that this name appears in the template file as well. These values should
+# match each other
+ENV_NAME=lxbe-env
+
+# As conda environment yml format does not seem to allow variables and we want
+# to rely on settings.sh to specify package versions, we programatically fill up
+# the env file with proper versions.
+cat $ENV_TEMPLATE > $ENV_FILE
+echo "  - conda=$CONDA_VERSION" >> $ENV_FILE
+echo "  - python=$PYTHON_VERSION" >> $ENV_FILE
+echo "  - openocd=$OPENOCD_VERSION" >> $ENV_FILE
```

Comment:
I think it makes sense, parsing it will be trivial and will allow us to drop this quasi-auto-generation of yml. It's a shame conda does not allow variables substitution. Even python-pip does.

Reply:
@mithro So we gave some thought to it and this calls for more discussion.

- download-env.sh installs packages on per-need basis
- build-pack.sh installs all the packages at once
- the versions are the same in both scenarios
- some packages are downloaded via pip, some are via conda.

Right now cmake is the only pip-installed package that has a version specified. It turns out that Conda has it in an older, but still suitable version available.

This means we could have multiple requirements.txt files in scripts/dependencies. There could be requirements.txt for common stuff, requirements-zephyr.txt, requirements-tinyfpga_b2.txt etc.

This way we could:
a) have a clear distinction over what requires which packages
b) install via ``pip install -r`` in both download-env and build-pack.

For Conda we can do it similarly, and have multiple requirement files for different dependencies, and then use them all to create a pack.
There are, allegedly, some problems with using files as source for ``conda install``, but I'm sure it's manageable. The format is also not really specified, but I think that ``pip`` files work fine, just as yml files.

Parsing files is something we'd prefer to avoid, for example to be able to easily switch the format/backend in the future.

Summing up, the flow would be:

```
download-env.sh:
if zephyr then ensure_deps_for(zephyr)
if sth_else then ensure_deps_for(sth_else)

build-pack.sh:
ensure_all_deps()

ensure_deps_for(x):
    // in future: support docker/local installations
    conda install -f conda_deps_for_$x
    pip install -r pip_deps_for_$x

ensure_all_deps():
   for x in deps:
      ensure_deps_for(x)
```

What's your opinion on that?

Reply:
I like the idea of having;
 * `requirements-<board>.txt`
 * `requirements-<firmware>.txt`
 * `requirements-<cpu>.txt`

I'm unclear if it should be separate firmware / cpu files or if it should be `requirements-<firmware>-<cpu>`?

Another option is that we could use dummy conda packages with the requirements as dependencies?

Reply:
I believe I will be enough to have `requirements-<something>.txt`, with only one dimension. If more is needed, we can always add it.